float32 weight
point loc
float32 angle
float32[] ranges
line moveline